### PR TITLE
API CMS Implementation of RFC-1 Asset Abstraction (DataObject Refactoring) 

### PR DESCRIPTION
**Pull request by tractorcow**
Support new asset abstraction and remove filesystem sync

Necessary pull request companion to https://github.com/silverstripe/silverstripe-framework/pull/4652

### CODE REVIEW

File: code/controllers/AssetAdmin.php [hunk 1]
```diff
@@ -62,13 +58,11 @@ public function currentPageID() {
...
 
 	/**
-	 * Set up the controller, in particular, re-sync the File database with the assets folder./
+	 * Set up the controller
 	 */
 	public function init() {
 		parent::init();
 
-		// Create base folder if it doesnt exist already
-		if(!file_exists(ASSETS_PATH)) Filesystem::makeFolder(ASSETS_PATH);
```
Comment (chillu) [inline code/controllers/AssetAdmin.php]: Confirmed that flysystem creates `ASSETS_PATH` automatically: https://github.com/thephpleague/flysystem/blob/master/src/Adapter/Local.php#L73

Comment (tractorcow) [inline code/controllers/AssetAdmin.php]: If it's using that adapter, it does. :)

**Comment by tractorcow** [comment]
Since file link tracking was broken in the prior framework merge, this pull request will rely on https://github.com/silverstripe/silverstripe-framework/pull/4679 before it can pass tests.

File::updateLinks has been moved to be invoked after write, and link rewriting now uses the same mechanism as HtmlEditorField, which is functionality refactored out of that class into `Image:: regenerate_html_links`.

**Comment by tractorcow** [comment]
@sminnee would you like to review my implementation of ErrorPage logging? Hopefully I've built on top of your work in the most appropriate way. :)

**Comment by chillu** [comment]
I've merged the underlying framework PR and gave the tests a kick

**Comment by tractorcow** [comment]
Hope composer picks up the change straight away.

**Comment by tractorcow** [comment]
Damn it, something broke again. =(

**Comment by tractorcow** [comment]
fulltextsearch components (obviously not those tested in framework tests) needed fixing. See https://github.com/silverstripe/silverstripe-framework/pull/4682 for remaining blocker for this pull request.

**Comment by sminnee** [comment]
The MySQL build passed; re-running the others now. Fingers crossed!

**Comment by sminnee** [comment]
Looks like SQLite, PostgreSQL, and the Behat extension still have issues.

**Comment by tractorcow** [comment]
Yikes, it looks like behat extension needs updating for master.

**Comment by tractorcow** [comment]
I'm getting really fed up with these postgres errors. =(

**Comment by tractorcow** [comment]
Ok, so here's the problem:

SiteTreeBacklinksTest has hard coded file ids, which postgres doesn't like (since it doesn't reset the increment id in PostgreSQLDatabase::clearTable).

I tried changing it to use truncate, and it fixed that problem, but now there's a primary key violation error in the tests.

**Comment by tractorcow** [comment]
HMMM postgres is passing finally, but a minor behat failure is going on still. =(

**Comment by tractorcow** [comment]
I'm running this on `develop` branch temporarily so that I can extract the travis screenshots and debug this issue.

**Comment by tractorcow** [comment]
Tests are reporting success, but if you inspect the behat log you'll see that one of them failed.

**Comment by tractorcow** [comment]
The test passes running behat locally.

I think we should just probably merge this.

**Comment by tractorcow** [comment]
I've merged develop into master of behat-extension and removed the reference in this pull request.

**Comment by tractorcow** [comment]
That broken behat test is fixed. ;)

**Pull request merged**